Add custom attachment factory guide for XML (#3543)
GetStream/stream-chat-android: docusaurus/docs/Android/03-ui/06-guides/08-adding-custom-attachments.mdx

# Adding Custom Attachments

## Introduction

By default Stream Chat supports several built-in attachment types like image, video, file and giphy. You can also add your own types of attachments such as location, contact, audio, sticker, etc.

In this guide, we'll demonstrate how to build a date sharing feature. Chat users will be able to pick a date from a calendar, preview their selection in the message input and see the sent attachment within a message in the message list.

> This involves doing the following steps:

Reviewer: Just a proposal, not a mandatory change 🙂
`The following steps are supposed to be made to achieve the goal above:`

Reviewer 2: I think the general guideline we follow with writing docs is to use as few words as possible, even if it means we ditch certain formalities.

It took a while for me to get used to it well, but it works because people inherently lose interest when they are faced with more text.

Reviewer 3: Yeah the shorter and more direct the better. @kanat @MarinTolic @TGazica @bychkovdmitry 

People don't like reading much and want to get quick information. Here are a few examples:

"You can use this new feature to be able to build your own custom attachment factories." This explains things, but is very verbose. Instead, say:

"Use this new feature to build your custom attachment factories."

Much more straightforward, clear and concise, while conveying the same information.

> This involves doing the following steps:

Reviewer 3: Yours is a bit nicer I'd say

> 1. Customizing the message input in a way that it is capable of sending messages with `date` attachments.

Reviewer 2: ```suggestion
1. Customizing the message input so that it is capable of sending messages with `date` attachments.
```

2. Adding support for `date` attachments in the message list.

:::note
In this guide, we'll show only the main points concerning custom attachments. Smaller parts will be omitted for the sake of being concise.

You can find the full code from this guide on [GitHub](https://github.com/GetStream/Android-Samples/tree/main/custom-attachments).
:::

## Sending Date Attachments

> To begin with, you'll need to customize [MessageInputView](../04-message-components/04-message-input.mdx) so that the attachment button allows the user to pick dates:

Reviewer: First of all, you'll need to customize [MessageInputView](../04-message-components/04-message-input.mdx) to let a user pick a date through the attachment button:

Reviewer 3: This is a great example of above of being concise! :]

I'd just fix one typ:

First of all, you'll need to customize [MessageInputView](https://github.com/GetStream/stream-chat-android/04-message-components/04-message-input.mdx) to let the user pick a date using the attachment button:

Since the user is not doing something through a button, rather "with/using" it

> To begin with, you'll need to customize [MessageInputView](../04-message-components/04-message-input.mdx) so that the attachment button allows the user to pick dates:

Reviewer 2: I think Kanat's version makes slightly more sense, maybe we can even drop the `of all`.
Here's Kanat's version as a suggestion for easier adoption.

```suggestion
First of all, you'll need to customize [MessageInputView](../04-message-components/04-message-input.mdx) to let a user pick a date through the attachment button:
```

Reviewer 3: Agreed but I'd use this:

`First of all, you'll need to customize MessageInputView to let the user pick a date using the attachment button:`

```kotlin

> messageInputView.setAttachmentButtonClickListener { 

Reviewer 2: I would add comments here so that it's easier to scan.
Such as:

// Create a date picker
// Add an attachment to the message input when the user selects a date
// Show the date picker

The things being done here are standard Android so I would consider this optional, however sometimes it helps to make the code scannable, otherwise, people can get disinterested.

It's up to you whether you think this is a good suggestion or unnecessary.

Reviewer 3: Sounds good yeah!

    val dialogPickerDialog = MaterialDatePicker.Builder
        .datePicker()
        .build()

    dialogPickerDialog.addOnPositiveButtonClickListener {
        val date = DateFormat
            .getDateInstance(DateFormat.LONG)
            .format(Date(it))
        val attachment = Attachment(
            type = "date",
            extraData = mutableMapOf("payload" to date)
        )
        messageInputView.submitCustomAttachments(
            attachments = listOf(attachment),
            viewHolderFactory = DateAttachmentPreviewFactory()
        )
    }

    dialogPickerDialog.show(requireActivity().supportFragmentManager, null)
}
```

> To keep it simple we are reusing the existing attachment button to show the date picker dialog.

Reviewer 2: I would change the attachment button icon here. 
It changed its purpose, it should change its appearance as well. Especially since it's in the screenshot.

    override fun createAttachmentViewHolder(
        attachments: List<Attachment>,
        parent: ViewGroup,
    ): BaseSelectedCustomAttachmentViewHolder {
        return ItemDateAttachmentPreviewBinding
            .inflate(LayoutInflater.from(parent.context), parent, false)
            .let(::DateAttachmentPreviewViewHolder)
    }

    class DateAttachmentPreviewViewHolder(
        private val binding: ItemDateAttachmentPreviewBinding,
    ) : BaseSelectedCustomAttachmentViewHolder(binding.root) {

        override fun bind(attachment: Attachment, onAttachmentCancelled: (Attachment) -> Unit) {
            binding.dateTextView.text = attachment.extraData["payload"].toString()
            binding.deleteButton.setOnClickListener {
                onAttachmentCancelled(attachment)
            }
        }
    }
}
```

Now you can send a custom attachment of type `date` to the chat. The resulting UI will look like this:

![Sending Custom Attachment](../../assets/sending_custom_attachments.png)

Next, you'll need to build a custom attachment factory to render the item in the message list.

## Rendering Date Attachments

> To render a custom attachment in the [MessageListView](../04-message-components/02-message-list.mdx) you'll have to implement `AttachmentFactory` interface:

Reviewer 2: ```suggestion
To render a custom attachment in the [MessageListView](../04-message-components/02-message-list.mdx) you'll have to implement the `AttachmentFactory` interface:
```

```kotlin
interface AttachmentFactory {
    
    fun canHandle(message: Message): Boolean
    
    fun createViewHolder(
        message: Message,
        listeners: MessageListListenerContainer?,
        parent: ViewGroup,
    ): InnerAttachmentViewHolder
}
```

There are two methods that need to be implemented:

* `canHandle`: Checks whether the factory can handle the given attachments.

> * ```

Reviewer 2: This part is broken, there's a rogue code segment opened here.

```suggestion
```

* `createViewHolder`: Represents the attachment UI within the `MessageListView`.

Let's see how to create an `AttachmentFactory` that is capable of handling date attachments:

```kotlin

> class DateAttachmentFactory : AttachmentFactory {

Reviewer 2: I think we are missing comments here.
It's easy for the reader to get paralyzed by a lot of code here, I think it would be more clear if we also provided short descriptions of what the various methods achieve.

    ): InnerAttachmentViewHolder {
        return ItemDateAttachmentBinding
            .inflate(LayoutInflater.from(parent.context), parent, false)
            .let { DateAttachmentViewHolder(it, listeners) }
    }

    class DateAttachmentViewHolder(
        private val binding: ItemDateAttachmentBinding,
        listeners: MessageListListenerContainer?,
    ) : InnerAttachmentViewHolder(binding.root) {

        private lateinit var message: Message

        init {
            binding.dateTextView.setOnClickListener {
                listeners?.attachmentClickListener?.onAttachmentClick(
                    message,
                    message.attachments.first()
                )
            }
            binding.dateTextView.setOnLongClickListener {
                listeners?.messageLongClickListener?.onMessageLongClick(message)
                true
            }
        }

        override fun onBindViewHolder(message: Message) {
            this.message = message

            binding.dateTextView.text = message.attachments
                .first { it.type == "date" }
                .extraData["payload"]
                .toString()
        }
    }
}
```

> What's important here is how you're able to fetch the custom attachment data, using `attachment.extraData["payload"]`. You can format the data in any way you want here, which makes our attachments very powerful.

Reviewer: What's really exciting here is the way the custom attachment data is being fetched, using `attachment.extraData["payload"]`. You can format the data in any way you want here, which makes our attachments very powerful.

> What's important here is how you're able to fetch the custom attachment data, using `attachment.extraData["payload"]`. You can format the data in any way you want here, which makes our attachments very powerful.

Reviewer 2: ```suggestion
The important thing is the ability to fetch custom attachment data using `attachment.extraData["payload"]`. You can format the data in any way you want here, which makes our attachments very powerful.
```

I'm also unsure about the second sentence in the suggestion. Are we trying to convey the point that they can put different kinds of data inside `attachment.extraData`?

Reviewer 3: Yeah the 2nd point is that there is a lot of control over the data, so you can do whatever suits your use cases

To complete the date sharing feature you just need to provide `DateAttachmentFactory` via `ChatUI`:

```kotlin
ChatUI.attachmentFactoryManager = AttachmentFactoryManager(listOf(DateAttachmentFactory()))
```

> Date attachments should now be correctly rendered in the message input and in the message list like on the screenshot below:

Reviewer: Date attachments should be correctly rendered now in the message input and in the message list like on the screenshot below:

> * ```

Reviewer 2: Also a case of rogue code blocks

```suggestion
```

:::

## Sending Date Attachments

First of all, you'll need to customize [MessageInputView](../04-message-components/04-message-input.mdx) to let the user pick a date using the attachment button:

```xml
<!-- Customize the message input to show a custom icon for the attachment button -->
<io.getstream.chat.android.ui.message.input.MessageInputView
    android:id="@+id/messageInputView"
    android:layout_width="match_parent"
    android:layout_height="wrap_content"
    app:streamUiAttachButtonIcon="@drawable/ic_calendar" />
```

```kotlin
// Create an instance of a date picker dialog
val datePickerDialog = MaterialDatePicker.Builder
    .datePicker()
    .build()

// Add an attachment to the message input when the user selects a date
datePickerDialog.addOnPositiveButtonClickListener {
    val date = DateFormat
        .getDateInstance(DateFormat.LONG)
        .format(Date(it))
    val attachment = Attachment(
        type = "date",
        extraData = mutableMapOf("payload" to date)
    )
    messageInputView.submitCustomAttachments(
        attachments = listOf(attachment),
        viewHolderFactory = DateAttachmentPreviewFactory()
    )
}

> // Show the date picker dialog on a click on the attachment button

Reviewer 2: ```suggestion
// Show the date picker dialog when the attachment button is clicked
```
